rclrs/src/context/builder.rs

```diff
@@ -0,0 +1,148 @@
+    /// assert!(ContextBuilder::new(invalid_remapping).build().is_err());
+    /// let valid_remapping = ["--ros-args", "--remap", "__node:=my_node"].map(String::from);
+    /// assert!(ContextBuilder::new(valid_remapping).build().is_ok());
+    /// ```
+    pub fn new(args: impl IntoIterator<Item = String>) -> ContextBuilder {
+        let mut domain_id = 0;
+        // SAFETY: Getting the default domain ID, based on the environment
+        let ret = unsafe { rcl_get_default_domain_id(&mut domain_id) };
+        debug_assert_eq!(ret, 0);
+
+        ContextBuilder {
+            arguments: args.into_iter().collect(),
+            domain_id,
+        }
+    }
+
+    /// Sets the context domain id.
+    ///
+    /// The domain ID controls which nodes can send messages to each other, see the [ROS 2 concept article][1].
+    ///
+    /// [1]: https://docs.ros.org/en/rolling/Concepts/About-Domain-ID.html
+    pub fn domain_id(mut self, domain_id: usize) -> Self {
+        #[cfg(ros_distro = "foxy")]
+        std::env::set_var("ROS_DOMAIN_ID", domain_id.to_string());
+
+        self.domain_id = domain_id;
+        self
+    }
+
+    /// Builds the context instance and `rcl_init_options` in order to initialize rcl
+    ///
+    /// For example usage, see the [`ContextBuilder`][1] docs.
+    ///
+    /// [1]: crate::ContextBuilder
+    pub fn build(&self) -> Result<Context, RclrsError> {
+        // SAFETY: Getting a zero-initialized value is always safe
+        let mut rcl_context = unsafe { rcl_get_zero_initialized_context() };
+        let cstring_args: Vec<CString> = self
+            .arguments
+            .iter()
+            .map(|arg| {
+                CString::new(arg.as_str()).map_err(|err| RclrsError::StringContainsNul {
+                    err,
+                    s: arg.clone(),
+                })
+            })
+            .collect::<Result<_, _>>()?;
+        // Vector of pointers into cstring_args
+        let c_args: Vec<*const c_char> = cstring_args.iter().map(|arg| arg.as_ptr()).collect();
+        let rcl_init_options = self.create_rcl_init_options()?;
+        unsafe {
+            // SAFETY: This function does not store the ephemeral init_options and c_args
+            // pointers. Passing in a zero-initialized rcl_context is expected.
+            rcl_init(
+                c_args.len() as i32,
+                if c_args.is_empty() {
+                    std::ptr::null()
+                } else {
+                    c_args.as_ptr()
+                },
+                &rcl_init_options,
+                &mut rcl_context,
+            )
+            .ok()?;
+        }
+        Ok(Context {
+            rcl_context_mtx: Arc::new(Mutex::new(rcl_context)),
+        })
+    }
+
+    /// Creates a rcl_init_options_t struct from this builder.
+    ///
+    /// domain id validation is performed in this method.
+    fn create_rcl_init_options(&self) -> Result<rcl_init_options_t, RclrsError> {
```

Comment:
Personally, I would make the builder have a field of `rcl_init_options_t` type, instead of only building it in `build()`. But I don't have a good argument for it, so I'll leave it up to you whether you want to do that :)

Reply:
I accept your suggestion. That field give more flexible features to assign allocators and a domain_id.